tests/dates/tests.py

```diff
@@ -121,3 +123,34 @@ def test_dates_fails_when_given_invalid_order_argument(self):
             "year",
             order="bad order",
         )
+
+    @override_settings(USE_TZ=False)
+    def test_dates_trunc_datetime_fields(self):
+        pub_datetimes = [
+            datetime.datetime(2015, 10, 21, 18, 1),
+            datetime.datetime(2015, 10, 21, 18, 2),
+            datetime.datetime(2015, 10, 22, 18, 1),
+            datetime.datetime(2015, 10, 22, 18, 2),
+        ]
+        for pub_datetime in pub_datetimes:
+            Article(
+                pub_date=pub_datetime.date(),
+                pub_datetime=pub_datetime
+            ).save()
+        self.assertQuerysetEqual(
+            Article.objects.dates('pub_datetime', 'day', order='ASC'),
+            ["datetime.date(2015, 10, 21)",
+             "datetime.date(2015, 10, 22)"]
+        )
+
+    @skipUnless(connection.vendor == 'mysql', "Test valid only for MySQL")
+    def test_dates_avoid_datetime_cast(self):
+        Article(pub_date=datetime.date(2015, 10, 21)).save()
+        kinds = [
+            'day',
+            'month',
+            'year',
+        ]
+        for kind in kinds:
+            qs = Article.objects.dates('pub_date', kind)
+            self.assertNotIn(' AS DATETIME', str(qs.query))
```

Comment:
I prefer checking for the expected syntax since "NotIn" can fail in all sorts of ways (for example, if you have a typo in the string).